### PR TITLE
LVIIICNET-202 Refactored donations

### PR DESCRIPTION
**Pull request by mod1fy**
## Core changes
- used enums instead of string values
- moved liqpay checkout generation to liqpay service ( reuse purposes)
- classes for liqpay data and signature 
- replaced 4+ parameters with dtos
- unit tests

## To discuss
- liqpay helper
- converting obj to query string
[link](https://github.com/ivanko-yuz/TheraLang/pull/228/files#diff-8bb81605b707db236f0a37e754c2ff74)

### CODE REVIEW

File: Common/Common.csproj [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
   <PropertyGroup>
     <TargetFramework>netcoreapp2.2</TargetFramework>
+    <LangVersion>8</LangVersion>
```
Comment (ivanko-yuz) [inline Common/Common.csproj]: yaaay

File: TheraLang.BLL/LiqPay/LiqPayHelper.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+using System.Linq;
+using TheraLang.BLL.DataTransferObjects.Donations;
+
+namespace TheraLang.BLL.LiqPay
+{
+    public static class LiqPayHelper
+    {
+        public static string ConvertToQueryString<T>(this T obj)
+        {
+            return string.Join("&", obj.GetType()
```
Comment (ivanko-yuz) [inline TheraLang.BLL/LiqPay/LiqPayHelper.cs]: what about using string builder here?

File: TheraLang.BLL/Services/AuthenticationService.cs [hunk 3]
```diff
@@ -50,19 +50,25 @@ public async Task<string> Authenticate(User user)
         }
 
         public async Task<AuthUser> GetAuthUserAsync()
+        {
+            return await TryGetAuthUser() ?? throw new Exception($"Error while getting user id from token");
+        }
+
+        public async Task<AuthUser> TryGetAuthUser()
```
Comment (ivanko-yuz) [inline TheraLang.BLL/Services/AuthenticationService.cs]: it's either not good naming or we should refactor this method to match `bool TryGetValue(id, our value)` signature
https://ayoungdeveloper.com/post/2017-03-28-using-the-tryget-pattern-in-csharp/

File: TheraLang.DAL/Entities/Donation.cs [hunk 4]
```diff
@@ -1,34 +1,45 @@
...
+using Common.Enums;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
 
 namespace TheraLang.DAL.Entities
 {
     public class Donation
     {
-        public int Id { get; set; }
+        public Guid Id { get; set; }
```
Comment (ivanko-yuz) [inline TheraLang.DAL/Entities/Donation.cs]: awesome thanks!

File: TheraLang.DAL/Entities/Donation.cs [hunk 5]
```diff
@@ -1,34 +1,45 @@
...
         public decimal Amount { get; set; }
 
         [JsonProperty("currency")]
-        public string Currency { get; set; }
+        [JsonConverter(typeof(StringEnumConverter))]
+        public LiqPayCurrency Currency { get; set; }
 
         [JsonProperty("payment_id")]
-        public int PaymentId { get; set; }
+        public uint PaymentId { get; set; }
```
Comment (ivanko-yuz) [inline TheraLang.DAL/Entities/Donation.cs]: why?

File: TheraLang.Tests/Services/DonationServiceTests.cs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+                Signature = "unholN06qwKFFl/LB3tc1qegq4E="
+            };
+            
+            _liqPayInfoMock = new Mock<ILiqPayInfo>();
+            _liqPayInfoMock.Setup(l => l.PrivateKey).Returns("testPrivateKey");
+            _liqPayInfoMock.Setup(l => l.PublicKey).Returns("testPublicKey");
+
+            var repoMock = new Mock<IRepository<Donation>>();
+            repoMock.Setup(r => r.Get(It.IsAny<Expression<Func<Donation, bool>>>()))
```
Comment (ivanko-yuz) [inline TheraLang.Tests/Services/DonationServiceTests.cs]: do we know what values should be here? if so any we use no Any but the specific arugment?

File: TheraLang.Tests/Services/DonationServiceTests.cs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+            resultDonation.Should().NotBeNull();
+            resultDonation?.Amount.Should().BePositive().And.Be(actualDonationAmount);
+        }
+
+        private IEnumerable<Donation> GetTestData()
+        {
+            return new List<Donation>()
+            {
+                new Donation()
```
Comment (ivanko-yuz) [inline TheraLang.Tests/Services/DonationServiceTests.cs]: can we create some builders around this? not in scope of this PR

File: TheraLang.BLL/LiqPay/LiqPayHelper.cs (hunk 2 above)
Comment (mod1fy) [reply TheraLang.BLL/LiqPay/LiqPayHelper.cs]: string.Join() uses string builder

File: TheraLang.Tests/Services/DonationServiceTests.cs (hunk 6 above)
Comment (mod1fy) [reply TheraLang.Tests/Services/DonationServiceTests.cs]: This setup we could use in multiple tests
e.g passing valid/invalid expression, null

File: TheraLang.BLL/LiqPay/LiqPayHelper.cs (hunk 2 above)
Comment (ivanko-yuz) [reply TheraLang.BLL/LiqPay/LiqPayHelper.cs]: fair enough, also it's more readable and faster

File: TheraLang.BLL/Services/AuthenticationService.cs (hunk 3 above)
Comment (mod1fy) [reply TheraLang.BLL/Services/AuthenticationService.cs]: Cannot use out param in an async method so we decided to return value if it's valid and null if not

**Pull request merged**